File: WebScraping/scraper.py
import sys
import re
import os
import time
from pathlib import Path
from playwright.sync_api import sync_playwright

OUTPUT_DIR = Path(r"D:\Sentencias")

def sanitize_filename(text):
    if not text:
        return ""
    clean = re.sub(r'[\s\-/,.]+', '_', text.strip())
    clean = re.sub(r'[\\:*?"<>|]', '', clean)
    return clean

def main():
    global OUTPUT_DIR
    if not OUTPUT_DIR.exists():
        try:
            OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
            print(f"Creada la carpeta: {OUTPUT_DIR}")
        except Exception as e:
            print(f"Error creando {OUTPUT_DIR}: {e}")
            print(r"Guardando en el directorio de trabajo (C:\...\Output) en su lugar.")
            OUTPUT_DIR = Path("Output") 
            OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    max_successful = 10
    success_count = 0
    current_id = 1
    consecutive_failures = 0
    max_consecutive_failures = 20

    print("Iniciando web scraping...")
    
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(accept_downloads=True)
        page = context.new_page()
        
        while success_count < max_successful and consecutive_failures < max_consecutive_failures:
            url = f"https://sij.poderjudicial.gob.hn/sentences/{current_id}"
            print(f"==========================================")
            print(f"[{current_id}] Visitando {url} ...")
            
            try:
                page.goto(url, wait_until="networkidle")
                
                # Esperamos al elemento de título o al texto de error
                try:
                    page.wait_for_selector("div.col-md-9 h2 strong, p:has-text('Sentencia Consultada no Existe o no es Pública')", timeout=15000)
                except Exception as e:
                    print(f"[{current_id}] Timeout al cargar elementos. Saltando...")
                    consecutive_failures += 1
                    current_id += 1
                    continue
                
                # Verificar si no existe
                not_found = page.locator("p:has-text('Sentencia Consultada no Existe o no es Pública')").count()
                if not_found > 0:
                    print(f"[{current_id}] No existe. Saltando...")
                    consecutive_failures += 1
                    current_id += 1
                    continue
                
                consecutive_failures = 0 
                
                sentencia_raw = page.locator("div.col-md-9 h2 strong").inner_text().strip()
                
                # Función que busca el valor a partir de un texto en negrita
                def get_field_val(label_text):
                    try:
                        # Buscamos en todo el body
                        # En la app, las etiquetas y valores son celdas row col-sm-4 y col-sm-8
                        locators = page.locator(f"xpath=//strong[contains(text(), '{label_text}')]/ancestor::div[contains(@class, 'col-sm-4')]/following-sibling::div[contains(@class, 'col-sm-8')]")
                        if locators.count() > 0:
                            return locators.first.inner_text().strip()
                        return ""
                    except Exception as ex:
                        return ""

                fres_raw = get_field_val('Fecha de resolución')
                if not fres_raw:
                    fres_raw = get_field_val('Fecha resolución')

                magis_raw = get_field_val('Magistrado ponente')
                mate_raw = get_field_val('Materia')
                fsente_raw = get_field_val('Fecha de sentencia recurrida')
                
                # Limpiando variables
                def clean(part_str):
                    return sanitize_filename(part_str)

                sentencia_clean = clean(sentencia_raw)
                fres_clean = clean(fres_raw)
                magis_clean = clean(magis_raw)
                mate_clean = clean(mate_raw)
                fsente_clean = clean(fsente_raw)
                
                base_name = f"Sentencia_{sentencia_clean}"
                if fres_clean: base_name += f"_FRes_{fres_clean}"
                if magis_clean: base_name += f"_Magis_{magis_clean}"
                if mate_clean: base_name += f"_Mate_{mate_clean}"
                if fsente_clean: base_name += f"_FSente_{fsente_clean}"
                
                if len(base_name) > 200:
                    base_name = base_name[:200]
                
                pdf1_name = f"{base_name}.pdf"
                pdf2_name = f"{base_name}_TipoCertificadoPDF.pdf"
                
                pdf1_path = OUTPUT_DIR / pdf1_name
                pdf2_path = OUTPUT_DIR / pdf2_name
                
                # 1. GENERAR PRIMER PDF (Imprimir navegador)
                print(f"[{current_id}] Generando primera versión (Navegador): {pdf1_name}")
                page.pdf(path=str(pdf1_path), format="A4", print_background=True, scale=0.8)
                
                # 2. SEGUNDO PDF (Botón original de la página)
                # OJO: The original button might be an a href or a button. "IMPRIMIR" is a strong keyword.
                btn_imprimir = page.locator("button, a").filter(has_text=re.compile(r"^\s*IMPRIMIR\s*$", re.IGNORECASE))
                if btn_imprimir.count() > 0:
                    try:
                        print(f"[{current_id}] Botón IMPRIMIR existente. Iniciando descarga: {pdf2_name}")
                        with page.expect_download(timeout=45000) as download_info:
                            btn_imprimir.first.click()
                            # El subagent mencionó un popup "Confirmar" de tipo sweetalert con "Si, continuar"
                            btn_confirmar = page.locator("button:has-text('Si, continuar')")
                            try:
                                btn_confirmar.wait_for(state="visible", timeout=3000)
                                btn_confirmar.click()
                            except:
                                # Puede ser que se descargue sin confirmación en algunas páginas
                                pass
                                
                        download = download_info.value
                        download.save_as(str(pdf2_path))
                        print(f"[{current_id}] Descarga exitosa.")
                    except Exception as e:
                        print(f"[{current_id}] Falló la descarga del segundo documento: {e}")
                else:
                    print(f"[{current_id}] Botón IMPRIMIR no encontrado.")
                
                success_count += 1
                print(f"[{current_id}] -> PROGRESO: ({success_count}/{max_successful})")
                
            except Exception as e:
                print(f"[{current_id}] Error general al procesar la sentencia: {e}")
                consecutive_failures += 1
            
            # Pausa para no sobrecargar el servidor
            time.sleep(2)
            current_id += 1

        browser.close()
        print("Finalizado con éxito o por límite de fallas consecutivas.")

if __name__ == "__main__":
    main()

File: WebScraping/scraper.js
const fs = require('fs');
const path = require('path');
const { chromium } = require('playwright');

let OUTPUT_DIR = path.resolve('D:\\Sentencias');

function sanitizeFilename(text) {
    if (!text) return '';
    let clean = text.trim().replace(/[\s\-\/,.]+/g, '_');
    clean = clean.replace(/[\\:*?"<>|]/g, '');
    return clean;
}

(async () => {
    if (!fs.existsSync(OUTPUT_DIR)) {
        try {
            fs.mkdirSync(OUTPUT_DIR, { recursive: true });
        } catch (e) {
            OUTPUT_DIR = path.resolve('Output');
            if (!fs.existsSync(OUTPUT_DIR)) fs.mkdirSync(OUTPUT_DIR, { recursive: true });
        }
    }

    const start_id = parseInt(process.argv[2], 10) || 1;
    const end_id = parseInt(process.argv[3], 10) || 1000000;
    
    let current_id = start_id;
    let consecutive_failures = 0;
    const max_consecutive_failures = 40; // Mayor tolerancia por si el chunk cayó en una zona vacía grande

    console.log(`[Agente ${start_id}-${end_id}] Iniciando chunk...`);

    const browser = await chromium.launch({ headless: true });
    const context = await browser.newContext({ acceptDownloads: true });
    
    // Inject the Blob interception correctly before any navigation
    await context.addInitScript(() => {
        const originalCreate = URL.createObjectURL;
        window.__capturedBlobs = [];
        URL.createObjectURL = function(obj) {
            const url = originalCreate(obj);
            window.__capturedBlobs.push(url);
            return url;
        };
    });

    const page = await context.newPage();

    while (current_id <= end_id && consecutive_failures < max_consecutive_failures) {
        const url = `https://sij.poderjudicial.gob.hn/sentences/${current_id}`;
        
        // Verificar si ya fue procesado para retomar desde donde se quedó
        const existingFiles = fs.readdirSync(OUTPUT_DIR);
        if (existingFiles.some(f => f.includes(`_pag${current_id}.pdf`) && !f.includes(`TipoCertificadoPDF`))) {
            console.log(`[Agente ${start_id}-${end_id}] ID ${current_id} ya existe en disco. Omitiendo...`);
            current_id++;
            continue;
        }

        console.log(`==========================================`);
        console.log(`[Agente ${start_id}-${end_id}] Visitando ${url} ...`);

        try {
            await page.goto(url, { waitUntil: 'networkidle', timeout: 30000 });
            await page.evaluate(() => { window.__capturedBlobs = []; });

            try {
                await page.waitForSelector("h2 strong, p:has-text('Sentencia Consultada no Existe o no es Pública')", { timeout: 15000 });
            } catch (e) {
                console.log(`[Agente ${start_id}-${end_id}] Timeout al cargar elementos en ID ${current_id}. Saltando...`);
                consecutive_failures++;
                current_id++;
                continue;
            }

            const notFoundCount = await page.locator("p:has-text('Sentencia Consultada no Existe o no es Pública')").count();
            if (notFoundCount > 0) {
                console.log(`[Agente ${start_id}-${end_id}] ID ${current_id} no existe. Saltando...`);
                consecutive_failures++;
                current_id++;
                continue;
            }

            consecutive_failures = 0;
            const sentenciaRaw = await page.locator("h2 strong").innerText();

            async function getFieldVal(labelText) {
                try {
                    const loc = page.locator(`xpath=//div[contains(@class, 'font-weight-bold') and contains(text(), '${labelText}')]/following-sibling::div/p`);
                    if (await loc.count() > 0) {
                        return (await loc.first().innerText()).trim();
                    }
                    return "";
                } catch (e) {
                    return "";
                }
            }

            let fresRaw = await getFieldVal('Fecha de resolución');
            if (!fresRaw) fresRaw = await getFieldVal('Fecha resolución');
            
            const magisRaw = await getFieldVal('Magistrado ponente');
            const mateRaw = await getFieldVal('Materia');
            const fsenteRaw = await getFieldVal('Fecha de sentencia recurrida');

            const sentenciaClean = sanitizeFilename(sentenciaRaw);
            const fresClean = sanitizeFilename(fresRaw);
            const magisClean = sanitizeFilename(magisRaw);
            const mateClean = sanitizeFilename(mateRaw);
            const fsenteClean = sanitizeFilename(fsenteRaw);

            let baseName = `Sentencia_${sentenciaClean}`;
            if (fresClean) baseName += `_FRes_${fresClean}`;
            if (magisClean) baseName += `_Magis_${magisClean}`;
            if (mateClean) baseName += `_Mate_${mateClean}`;
            if (fsenteClean) baseName += `_FSente_${fsenteClean}`;

            if (baseName.length > 200) {
                baseName = baseName.substring(0, 200);
            }

            const pdf1Name = `${baseName}_pag${current_id}.pdf`;
            const pdf2Name = `${baseName}_TipoCertificadoPDF_pag${current_id}.pdf`;
            const pdf1Path = path.join(OUTPUT_DIR, pdf1Name);
            const pdf2Path = path.join(OUTPUT_DIR, pdf2Name);

            // ====== Capturar el TipoCertificadoPDF nativo usando el boton =======
            const btnImprimir = page.locator("button, a").filter({ hasText: /^\s*IMPRIMIR\s*$/i });
            if (await btnImprimir.count() > 0) {
                await btnImprimir.first().click();
                try {
                    await page.locator("button:has-text('Si, continuar')").waitFor({ state: 'visible', timeout: 3000 });
                    await page.locator("button:has-text('Si, continuar')").click();
                } catch (e) {}

                // Waits for html2canvas buffer
                await page.waitForTimeout(6000);

                const blobs = await page.evaluate(() => window.__capturedBlobs);
                if (blobs && blobs.length > 0) {
                    const finalUrl = blobs[blobs.length - 1];
                    const base64String = await page.evaluate(async (blobUrl) => {
                        try {
                            const res = await fetch(blobUrl);
                            const blob = await res.blob();
                            return new Promise((resolve) => {
                                const reader = new FileReader();
                                reader.onloadend = () => resolve(reader.result.split(',')[1]);
                                reader.readAsDataURL(blob);
                            });
                        } catch(err) {
                            return null;
                        }
                    }, finalUrl);
                    
                    if (base64String) {
                        fs.writeFileSync(pdf2Path, Buffer.from(base64String, 'base64'));
                        console.log(`[Agente ${start_id}-${end_id}] ID ${current_id}: Certificado nativo descargado.`);
                    }
                }
            }

            // ====== Generar PDF 1 (METADATOS ONLY) ====
            await page.evaluate(() => {
                const selectors = ['app-header', '#footer', 'app-menu-side-bar', '.page-header', '#sentenceTextDiv', '.accordion', '.btn-modern'];
                selectors.forEach(selector => {
                    document.querySelectorAll(selector).forEach(el => el.remove());
                });
                document.querySelectorAll('span').forEach(span => {
                    if (span.innerText && span.innerText.includes('Nota: Si tiene problemas')) {
                        span.remove();
                    }
                });
            });

            console.log(`[Agente ${start_id}-${end_id}] ID ${current_id}: Web descargado.`);
            await page.pdf({ path: pdf1Path, format: 'A4', printBackground: true, scale: 0.8 });

        } catch (e) {
            console.log(`[Agente ${start_id}-${end_id}] Error en ID ${current_id}: ${e.message}`);
            consecutive_failures++;
        }

        current_id++;
    }

    await browser.close();
    console.log(`[Agente ${start_id}-${end_id}] Finalizado.`);
    
    // Si se salió por muchas fallos seguidos, mandar señal a maestro para dejar de spammear
    if (consecutive_failures >= max_consecutive_failures) {
        if (process.send) {
            process.send('STOP_MASTER');
        }
    }

})();
